lib/rspec/matchers/built_in/be_instance_of.rb

```diff
@@ -14,17 +14,21 @@ def description
       private
 
         def match(expected, actual)
-          if actual_object_respond_to?(actual, :instance_of?)
-            actual.instance_of?(expected)
+          result = nil
+          if executable? { result = actual.instance_of?(expected) }
```

Comment:
I am **personally** not a huge fan of using a block in that case. What about passing args to the method and simply rescue `NoMethodError`.
 
This a not a blocker to accept the PR. I am waiting for @JonRowe's final review. 🙌🏻

Reply:
Yeah, I thought about it, but it just seemed like I basically have to pass everything to that method. Because the method will end up being something like this:
```ruby
def executable?(object, method, *args)
  object.public_send(method, *args)
  true
rescue
  false
end
```

I don't know, I just liked that block trick better :)

Reply:
This can be simplified to:

```
def match(expected, actual)
  actual.instance_of?(expected)
rescue NoMethodError
  raise ::ArgumentError, "The #{matcher_name} matcher requires that " \
                         "the actual object responds to #instance_of? method " \
                         "but a `NoMethodError` was encountered instead."
end
```

The block trick adds execution time

Reply:
Remember we're just trying to surface a friendlier message to those objects not supporting a standard Ruby interface.

Reply:
@JonRowe  are we talking about **only** that case with `#instance_of?` check, right? So I'm leaving the block trick in `#is_a?`/`#kind_of?` check, right? Or it's still better to replace it with a method there?